### ISSUE TITLE
Wrong wrapping with contain width and text too long

### ISSUE THREAD

**Issue by pecoram** [lightning-js/renderer#268]
With `contain: "width"` a long text is not wrapped correctly:

![image](https://github.com/lightning-js/renderer/assets/33352553/f703db99-6c19-4ca4-a9d2-d8b973a2f9a2)

this issue is for both canvas and sdf

**Comment by marcel-danilewicz-consult-red** [comment]
Just to clarify - is `contain: "width"` equivalent of css `text-wrap: wrap`? If yes, then I think, that this behavior may be correct. I checked https://developer.mozilla.org/en-US/docs/Web/CSS/text-wrap and it behaves similarly. The text overflows only if it is continuous, without whitespaces.

**Comment by pecoram** [comment]
> Just to clarify - is `contain: "width"` equivalent of css `text-wrap: wrap`? If yes, then I think, that this behavior may be correct. I checked https://developer.mozilla.org/en-US/docs/Web/CSS/text-wrap and it behaves similarly. The text overflows only if it is continuous, without whitespaces.

`contain: "width"` must behave as `text-wrap: wrap` so if a word is too long the excess part must wrap. Just like the example on the page you linked

**Comment by marcel-danilewicz-consult-red** [comment]
I was able to achieve the same result as in lightning3:
![image](https://github.com/lightning-js/renderer/assets/172387148/84c98ad9-e163-423d-94a3-4fd9d7ef1201)

**Comment by marcel-danilewicz-consult-red** [comment]
In this case should this ticket be treated as invalid or maybe we want to implement the wrap-word functionality in it's scope?

**Comment by marcel-danilewicz-consult-red** [comment]
I tried tinkering with the code in case we wanted to implement this. Is this the expected behavior for "wrap-word" or should the overflow at the bottom not happen?
![image](https://github.com/lightning-js/renderer/assets/172387148/2a0006f0-e1e2-4237-bd76-67c22e397096)

**Comment by pecoram** [comment]
In my opinion it is wrong to wrap with spaces but if we want we can add a separate parameter that we call wordWrap and if it is true it truncates the word and not the last space

**Comment by marcel-danilewicz-consult-red** [comment]
Hi @wouterlucas 
I'd like to clarify the requirements for this functionality. I see two solutions:
a) any word, that exceeds the width limit should wrap
b) the word, that exceeds the width limit should be moved to the next line. If it stil exceeds the limit it should wrap until it fits
Is either one correct? Are there any other requirements?

**Comment by wouterlucas** [comment]
We should provide options instead of a `boolean`, because implementing an algorithm that tries to catch all scenarios gracefully will very likely perform very poorly as @elsassph pointed out.

We could have the dev point out what he wants explicitly.

My suggestion:
```
word-wrap: [space|wrap|pretty]
space - Breaks on spaces, as efficient as possible
wrap - Does a normal wrap, breaks on spaces if available if it still doesn't fit break on character
pretty - Like L2, breaks on space if possible, breaks on character but tries to fill/remove characters if needed to reduce orphans
```

Flex and wrap algo's can be detrimental to performance and hopefully `space` will work for the majority of the cases in latin languages without sacrificing too much. We could also start with space/wrap as a first phase?

**Comment by elsassph** [comment]
L2 had just an extra `wordBreak` in addition to regular wrapping options, and it was only breaking a single too-long word as in the linked issue. It seems sensible enough to me - only the implementation in L2 wasn't great.

The suggested `wrap` doesn't make sense to me - I don't imagine one would want systematically breaking words. This is scope/complexity creep IMHO.

**Comment by marcel-danilewicz-consult-red** [comment]
Maybe let's start with 
word-wrap: [space|wrap]

**Comment by elsassph** [comment]
Would be preferable to align with CSS rather than invent new names like "pretty" https://developer.mozilla.org/en-US/docs/Web/CSS/word-break

**Comment by marcel-danilewicz-consult-red** [comment]
Definitely. Also it is going to be a good reference.

**Comment by wouterlucas** [comment]
> Would be preferable to align with CSS rather than invent new names like "pretty" https://developer.mozilla.org/en-US/docs/Web/CSS/word-break

https://developer.mozilla.org/en-US/docs/Web/CSS/text-wrap#pretty 🙃 - however I agree that's more of a flex / text layout property over just word breaks. 

And the line between text layouting and word breaking is blurry, if we want to limit this to word break functionality only. I agree that [word-break](https://developer.mozilla.org/en-US/docs/Web/CSS/word-break) is a fine guidance. Likely limited to 
normal, break all or keep all?

**Comment by elsassph** [comment]
Weeeell the CSS pretty option has a fair explanation :D

For now I think we can support `normal` and `break` - and unlike what I suggested it seems that the breaking approach that @marcel-danilewicz-consult-red started would be the right thing to do. 

However `keep-all` option requires testing the unicode range of the characters, so it's a bit too much scope for a first shot.

**Comment by marcel-danilewicz-consult-red** [comment]
@elsassph so It's ok to keep current implementation and just change word-wrap to [normal | break]?

**Comment by elsassph** [comment]
I don't get to decide, I just a contributor 😅
I'd be reluctant to make an API change now that it's in v1, so it seems better to add an extra `wordBreak` option - but that would be following your current implementation.

**Comment by wouterlucas** [comment]
Agree, exposing `wordBreak` as TrProp is fine. Both `wordWrap` and `wordBreak` are already available internally (as booleans tho) and wrap is only flipped when contain is enabled.

**Comment by marcel-danilewicz-consult-red** [comment]
@wouterlucas I got a bit confused. Do we need both `wordWrap` and `wordBreak` exposed? Should they be boolean or string literals?

**Comment by wouterlucas** [comment]
Oh wait, the wordBreak isn't doing anything in https://github.com/lightning-js/renderer/blob/main/src/core/text-rendering/renderers/LightningTextTextureRenderer.ts#L74 as its always set to false and nothing seems to look at it down below. So that would need to hooked up/changed.

Also having two boolean's would be confusing as what do we do if they're both `true`. In that case a string with 
`normal` or `break` is safer.

Sorry for the confusion, the text rendering code is relatively new to me and the `wordBreak` option on the Canvas side threw me off.
